### PR TITLE
Allow headers size extend to maxRequestHeadersSize in http client

### LINKED ISSUE #12436: Allow headers size extend to maxRequestHeadersSize in http client
**Jetty version(s)**
Jetty Version 10-12

**Enhancement Description**

In HttpSenderOverHTTP and HttpClient, there is a "headerBuffer" which allocated by the "httpclient.getRequestBufferSize".
If the buffer overflowed,  there will be an IllegalArgumentException("Request header too large").  To support higher headers, for examples 32K or 64K,  has to set the request buffer size higher accordingly right now.

But 99% use cases, we don't need such high headerBuffer.  Also when backend connections stuck,  these high buffer will bump up the memory consumption significantly.

The enhancement will be have a new limit for headers,  it will be  "httpclient.maxRequestHeadersSize".   The headerBuffer still starts will the "httpclient.getRequestBufferSize".  While the headerBuffer runs into overflow state.  The headerBuffer will be allocated as "httpclient.maxRequestHeadersSize", and continue the flow.

In this case, 99% use cases can be covered by smaller header buffer, and 1% cases run into buffer extension.

### PR DESCRIPTION
**Pull request by shaoxt**
Fixes #12436

Allow headers size extend to maxRequestHeadersSize in http client

### CODE REVIEW

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java [hunk 1]
```diff
@@ -135,6 +135,7 @@ public class HttpClient extends ContainerLifeCycle implements AutoCloseable
     private String defaultRequestContentType = "application/octet-stream";
     private boolean useInputDirectByteBuffers = true;
     private boolean useOutputDirectByteBuffers = true;
+    private int maxRequestHeadersSize = 32 * 1024;
```
Comment (sbordet) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: This configuration is HTTP/1.1 specific.
Please move it to `HttpClientTransportOverHTTP`.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java [hunk 2]
```diff
@@ -1156,4 +1157,14 @@ public interface Aware
     {
         void setHttpClient(HttpClient httpClient);
     }
+
+    public int getMaxRequestHeadersSize()
```
Comment (sbordet) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: After moving this method to `HttpClientTransportOverHTTP`, please annotate it with:

```
    @ManagedAttribute("The maximum allowed size in bytes for the HTTP request headers")
    public int getMaxRequestHeadersSize()
```

**Review by sbordet — CHANGES_REQUESTED** [review]
Can you please:

1. Apply the review comments.
1. Rebase these changes against branch `jetty-12.1.x`, and squash them in just 1 commit.
2. Move the tests to `HttpClientTest` and actually call `setMaxRequestHeadersSize()` in the test so that we will test it is working.

I am probably make your changes part of a larger set of changes that improve the behavior of the current implementation.

Thanks!

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java (hunk 1 above)
Comment (shaoxt) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: Thanks, moved.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java (hunk 2 above)
Comment (shaoxt) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: Thanks, moved.

**Comment by shaoxt** [comment]
Thanks for reviewing the PR.    Created a new PR for required changes 
https://github.com/jetty/jetty.project/pull/12499
For some reasons, the last commits I tried to squash them in 1 commit. but git said the commits from two different parents.

> Can you please:
> 
> 1. Apply the review comments.
> 2. Rebase these changes against branch `jetty-12.1.x`, and squash them in just 1 commit.
> 3. Move the tests to `HttpClientTest` and actually call `setMaxRequestHeadersSize()` in the test so that we will test it is working.
> 
> I am probably make your changes part of a larger set of changes that improve the behavior of the current implementation.
> 
> Thanks!

**Pull request closed without merging**